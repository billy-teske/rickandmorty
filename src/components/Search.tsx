import styles from './Search.module.css';

const Search = () => {
    // const handleSubmit = () => {
    //     alert('submit');
    // };

    return (
        <div className={styles.search}>
            <input type="text" placeholder="Buscar" aria-label="Buscar" className={styles.input} />
        </div>
    );
};

export default Search;
