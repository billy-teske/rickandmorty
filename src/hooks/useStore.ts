import { create } from 'zustand';
import { TCharacter } from '../api/getCharacter';

type TState = {
    characters: TCharacter[];
    setCharacters: (newItems: TCharacter[]) => void;
    page: number;
    nextPage: number;
    toPage: () => void;
    toNextPage: () => void;
};

const useStore = create<TState>((set) => ({
    characters: [],
    setCharacters: (newsCharacters) => set((state) => {
        const addedCharacters = newsCharacters.filter(
            (newCharacter) => !state.characters.find(
                character => newCharacter.id === character.id
            )
        );

        return {
            characters: [...state.characters, ...addedCharacters],
        };
    }),
    page: 0,
    toPage: () => set((state) => ({ page: state.nextPage })),
    nextPage: 1,
    toNextPage: () => set((state) => ({ nextPage: state.page + 1 })),
}));

export default useStore;
