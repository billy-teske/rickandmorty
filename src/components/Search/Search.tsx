import useStore from '../../hooks/useStore';
import styles from './Search.module.css';

const Search = () => {
    const { setFilter } = useStore(state => state);

    const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
        event.preventDefault();
    };

    const handlerOnChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        setFilter(event.target.value);
    };

    return (
        <form onSubmit={handleSubmit} className={styles.search} data-testid="form">
            <input
                type="search"
                placeholder="Search"
                aria-label="Search"
                data-testid="search"
                className={styles.input}
                onChange={handlerOnChange}
            />
        </form>
    );
};

export default Search;
