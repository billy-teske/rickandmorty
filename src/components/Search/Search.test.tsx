import { act, fireEvent, render, waitFor } from '@testing-library/react';
import useStore from '../../hooks/useStore';
import Search from './Search';

jest.mock('../../hooks/useStore', () => jest.fn());

const mockSetFilter = jest.fn();
(useStore as unknown as jest.Mock).mockImplementation((fn) => {
    fn();
    return {
        setFilter: mockSetFilter,
    };
});

describe('Search Component', () => {
    it('should render a footer', () => {
        const result = render(
            <Search />
        );

        expect(result).toMatchSnapshot();
    });

    it('should set filter when change search box', () => {
        const result = render(
            <Search />
        );

        const { getByTestId } = result;
        const inputSearch = getByTestId('search');
        act(() => {
            fireEvent.change(inputSearch, {
                target: {
                    value: 'a'
                }
            });
        });

        waitFor(() => {
            expect(mockSetFilter).toHaveBeenCalled();
        });
    });

    it('should ignore form event submit', () => {
        const mockPreventDefault = jest.fn();
        const result = render(
            <Search />
        );

        const { getByTestId } = result;
        const formSearch = getByTestId('form');

        act(() => {
            fireEvent.submit(formSearch, { preventDefault: mockPreventDefault });
        });

        waitFor(() => {
            expect(mockPreventDefault).toHaveBeenCalled();
        });
    });
});
