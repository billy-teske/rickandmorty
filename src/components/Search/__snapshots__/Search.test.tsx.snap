// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Component should render a footer 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="search"
        data-testid="form"
      >
        <input
          aria-label="Search"
          class="input"
          data-testid="search"
          placeholder="Search"
          type="search"
        />
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="search"
      data-testid="form"
    >
      <input
        aria-label="Search"
        class="input"
        data-testid="search"
        placeholder="Search"
        type="search"
      />
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
